man/man1/modysnap.1
.TH MODYSNAP 1NEMO "10 June 2011"
.SH NAME
modysnap \- convert Mody's bin pos-vel files to snapshot
.SH SYNOPSIS
\fBmodysnap\fP [parameter=value]
.SH DESCRIPTION
\fBmodysnap\fP converts snapshot-style datasets from
the \fIn-mody\fP MOND N-body integrator to a standard
NEMO \fIsnapshot(5NEMO)\fP.
.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBin=\fP
Input mody pos-vel file (moutXX.bin style). Multiple files can be given.
No default.
.TP
\fBout=\fP
Output snapshot file. No default.
.TP
\fBheadline=\fP
Random verbiage []     
.TP
\fBswap=\fP
Swap bytes in each int/float if endianism differs . This is only needed if
the input files come from a machine with the opposite endianism as your
local host.
Not implemented yet.
.TP
\fBheader=t|f\fP
Read the header?  Needed for MODY style, but e.g. QuMOND only carries
particle information, so nbody= is needed instead (see below)
Default: true
.TP
\fBnbody=\fP
Set the number of particles to read. Only used when \fPheader=f\fP for QuMOND
style files. In principle by setting smaller value of nbody, you could use 
this to split the output snapshot in pieces.
.SH EXAMPLES
Here is an example of running n-mody. First create in input file:
.nf
  % cp $NEMO/usr/n_mody/example_init/newgalaxy.data .
  % rungb
   ...
  20.013u 7.036s 0:27.05 99.9%	0+0k 0+0io 0pf+0w
  % modysnap mout00.bin snap0
   ...	nbody=2048000 mods=1
   ...	totms=1.000000 tnow=0.000000 tdyn=8.332943 vir=0.000000
.fi
then get the parameters for the integration and run the standard benchmark
.fi
  % cp $NEMO/usr/n_mody/run/input.data .
  % rmond
  565.048u 6.426s 9:31.90 99.9%	0+0k 0+0io 0pf+0w
.nf
.fi
though notice this benchmark does not produce extra mout01.bin etc. files (nout=0)
.PP
To process multiple files, a Unix pipe should be used:
.nf
  % cat mout??.bin | modysnap - mout.snap
.fi
.SH LIMITATION
Currently only the \fImoutXX.bin\fP (pos-vel) files can be processed. 
However, it's not hard to have a \fIpoutXX.bin\fP type to be merged
in and produce a snapshot with potentials as well.
.PP
There is no good alternative for the \fImondXX.bin\fP files in
NEMO yet.
.SH FILES
There are a number of files used by \fIrmond\fP:
.PP
\fBmoutXX.bin\fP: a small 5 integer and 5 real header, followed by nbody
float "posvel" vectors.
.PP
\fBmondXX.bin\fP: grid data
.PP
\fBpoutXX.bin\fP: particle potentials. Same header

.SH SEE ALSO
rungb, rmond, runmond(1NEMO), snapmody(1NEMO), snapshot(5NEMO)
.SH FILES
$NEMO/src/nbody/io/modysnap.c  -- code
.SH AUTHOR
Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
23-Feb-09	V1.0 Created	PJT
26-feb-09	V1.2 process multiple files if present	PJT
10-jun-11	V1.3 added header= and nbody=	PJT
.fi


src/nbody/io/modysnap.c
/*
 *  convert mody's 'pos-vel' dumps to snapshot - in single precision
 *
 *    23-feb-09  PJT  created ; 
 *    25-feb-09  PJT  added pin= (but not processing) , wrote out few more pars
 *    26-feb-09  PJT  process multiple MOUT files
 *    10-jun-11  PJT  also allow QuMOND bin files (header=f, nbody=...)
 */
 
#define SinglePrec
#define SINGLEPREC
#undef  DOUBLEPREC

#include <nemo.h>
#include <unfio.h>
#include <snapshot/snapshot.h>
#include <snapshot/body.h>

string defv[] = {
    "in=???\n       Input mody moutXX.bin pos-vel file",
    "out=???\n      Output snapshot file",
    "pin=\n         Input mbody poutXX.bin potential file (optional)",
    "swap=f\n       Swap bytes in each float if endianism differs",
    "header=t\n     Skip header (e.g. for QuMOND)",
    "nbody=0\n      Need this if header is absent",
    "headline=\n    Random verbiage",
    "VERSION=1.3\n  10-jun-2011 pjt",
    NULL,
};

string usage = "convert MODY's bin pos-vel files to snapshot";

extern void bswap(void *vdat, int len, int cnt);
 
void nemo_main()
{
    stream instr, outstr;
    real *fbuf, *mbuf, tsnap, mass0, mass1, tdyn;
    bool Qswap = getbparam("swap");
    bool Qhead = getbparam("header");
    long nread;
    int i, mond_ind, model;
    int nbody = getiparam("nbody");
    int cs = CSCode(Cartesian, NDIM, 2);
    int ipar[5];
    float rpar[5], posvel[6];

    if (!Qhead && nbody==0) 
      error("Need nbody= if you leave out header");
    

    if (hasvalue("headline"))
        set_headline(getparam("headline"));
    instr = stropen(getparam("in"),"r");
    outstr = stropen(getparam("out"),"w");
    put_history(outstr);

    if (hasvalue("pin")) 
      warning("pin=%s not yet processed",getparam("pin"));


    while (1) {
      if (Qhead) {
	nread = unfread(instr, (char *)ipar, 5*sizeof(int));
	if (nread == 0) break;
	dprintf(0,"ipar: nbody=%d mods=%d\n",
		ipar[0],ipar[1]);

	nread = unfread(instr, (char *)rpar, 5*sizeof(float));
	dprintf(0,"rpar: totms=%f tnow=%f tdyn=%f vir=%f\n",
		rpar[0],rpar[1],rpar[2],rpar[3]);

	nbody = ipar[0];
	model = ipar[1];
	mond_ind = ipar[2];
	/* ipar[3,4] still unused */
	
	mass0 = rpar[0];
	mass1 = mass0/nbody;
	tsnap = rpar[1];
	tdyn  = rpar[2];
	/* rpar[3,4] still unused */
      }

      /* fbuf holds the array of posvel's
       * mbuf holds the (all the same ) masses - not used in header=f mode
       */
      fbuf = (real *) allocate(nbody*6*sizeof(real)); 
      mbuf = (real *) allocate(nbody*sizeof(real));   

      for (i=0; i<nbody; i++) {
	mbuf[i] = mass1;
	dprintf(0,"P %d\n",i);
	nread = unfread(instr, (char *) &fbuf[i*6], 6*sizeof(real)); 
	if (!Qhead && nread==0) break;
	if (nread<1) error("Early EOF for i=%d (%d)",i,nread);
      }
      if (!Qhead && nread==0) {
	free(fbuf);
	free(mbuf);
	break;
      }
    
      /* write out snapshot for NEMO */

      put_set(outstr,SnapShotTag);
      put_set(outstr,ParametersTag);
       put_data(outstr, NobjTag, IntType, &nbody, 0);
       if (Qhead) {
	 put_data(outstr, TimeTag, RealType, &tsnap, 0);
	 put_data(outstr, "mass",      RealType, &mass0, 0);
	 put_data(outstr, "tdyn",      RealType, &tdyn, 0);
	 put_data(outstr, "model",     IntType, &model, 0);
	 put_data(outstr, "mond_ind",  IntType, &mond_ind, 0);
       }
      put_tes(outstr,ParametersTag);
      put_set(outstr,ParticlesTag);
        put_data(outstr,CoordSystemTag, IntType, &cs, 0);
	if (Qhead)
	  put_data(outstr,MassTag,RealType,mbuf,nbody,0);
        put_data(outstr,PhaseSpaceTag,RealType,fbuf,nbody,2,NDIM,0);
      put_tes(outstr,ParticlesTag);
      put_tes(outstr,SnapShotTag);

      free(fbuf);
      free(mbuf);
    }

    strclose(instr);
    strclose(outstr);
}




